Memory access error from invalid code

So, I made a mistake with my code, and accidentally wrote else instead of elsif, and had a condition after it on the same line, as in:

`else !var.only_once`

Instead of producing a compiler error, this produced an "Invalid memory access" error. I'm not sure if this is a serious issue, being that it only happened because I wrote invalid code, but better safe than sorry, right?

> Invalid memory access (signal 11) at address 0x8
[0x109ddb987] *CallStack::print_backtrace:Int32 +39
[0x1095235de] __crystal_sigfault_handler +46
[0x7fffb58ffb3a] _sigtramp +26
[0x10ab3d227] _ZN4llvm8Constant15getAllOnesValueEPNS_4TypeE +29
[0x10ab3d2e2] _ZN4llvm8Constant15getAllOnesValueEPNS_4TypeE +216 (2 times)
[0x10ab42ac7] _ZN4llvm12ConstantExpr6getNotEPNS_8ConstantE +17
[0x10ab2bfa9] _ZN4llvm9IRBuilderINS_14ConstantFolderENS_24IRBuilderDefaultInserterEE9CreateNotEPNS_5ValueERKNS_5TwineE +41
[0x10ab50192] LLVMBuildNot +39
[0x109a3144e] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10302
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a31860] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11344
[0x109a317fe] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11246
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a317fe] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11246
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a34e0a] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +25082 (2 times)
[0x109a30423] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6163
[0x109a2ff4c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +4924
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a50c94] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a50c94] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a4892d] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x109a5053d] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a4a4b9] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +313
[0x109a2f31c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1804
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a4892d] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x109a5053d] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a2fdb7] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +4519
[0x109a4a4b9] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +313
[0x109a2f31c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1804
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a4892d] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x109a5053d] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x109a502c6] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +3894
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a55194] *Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool) +1140
[0x109a503b6] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4134
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a4a4b9] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +313
[0x109a2f31c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1804
[0x109a34e0a] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +25082 (2 times)
[0x109a30423] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6163
[0x109a2ff4c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +4924
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a50c94] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a50c94] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +6404
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a30423] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6163
[0x109a317fe] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +11246
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a33c4e] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +20542
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a4892d] *Crystal::CodeGenVisitor#target_def_fun<Crystal::Def+, Crystal::Type+>:LLVM::Function +2445
[0x109a5053d] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4525
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a4a4b9] *Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil) +313
[0x109a2f31c] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1804
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x109a454e0] *Crystal::CodeGenVisitor#codegen_fun<String, Crystal::Def+, Crystal::Type+, Bool, Crystal::CodeGenVisitor::ModuleInfo, Bool, Bool>:LLVM::Function +4368
[0x109a2fdb7] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +4519
[0x109a50425] *Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool +4245
[0x109a314cc] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +10428
[0x109a30423] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +6163
[0x109a2f78e] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +2942
[0x109a2f3b4] *Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil +1956
[0x1099fb03a] *Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil) +490
[0x109d5ab34] *Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +116
[0x109daf109] *Crystal::Command#run_command<Bool>:Nil +153
[0x10953b893] *Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil) +20931
[0x109501864] main +212
make: *** [test] Error 11

--- a/spec/compiler/codegen/not_spec.cr
+++ b/spec/compiler/codegen/not_spec.cr
@@ -55,4 +55,16 @@ describe "Code gen: not" do
       !a.is_a?(String) && !a
       )).to_b.should be_false
   end
+
+  it "codegens not with inlinable value (#6451)" do
+    codegen(%(
+      class Test
+        def test
+          false
+        end
+      end
+
+      !Test.new.test
+      nil))
+  end
 end

--- a/src/compiler/crystal/codegen/codegen.cr
+++ b/src/compiler/crystal/codegen/codegen.cr
@@ -815,7 +815,9 @@ module Crystal
     end
 
     def visit(node : Not)
-      accept node.exp
+      request_value do
+        accept node.exp
+      end
       @last = codegen_cond node.exp.type.remove_indirection
       @last = not @last
       false

